### PR TITLE
Custom app patterns

### PR DESCRIPTION
**Pull request by dsbegnoche**
Ability to customize Patterns on a per app basis.

1.  Default patterns are created on App Create from appsettings.json
2.  Patterns can be modified in App Settings 
3. Custom patterns can set Name, Pattern and Default Message
4. Patterns must be unique in both Name and Pattern

Questions remaining on implementation:
1. I removed UI Controller in favor of AppPatterns controller. Patterns now behave much like clients/languages etc...  However, this does mean that existing apps will not have patterns showing.  What is the best way to handle these existing apps and getting default patterns created?  A migration process or another option?
2. We want to update any pattern in use on a field when it is updated via Settings.  So far, the only implementation we have of this is pretty ugly.  It requires looping through all the schemas to find any fields with the pattern applied, then update the field with the new value.  It seems their has to be a better way, but currently have not discovered that. Any ideas would be appreciated.   Additionally, is this behavior something that makes sense in Squidex or would it be best for updates to not touch the existing schemas?

### CODE REVIEW

File: src/Squidex.Domain.Apps.Core.Model/Apps/AppPattern.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿// ==========================================================================
+//  AppPattern.cs
+//  Squidex Headless CMS
+// ==========================================================================
+//  Copyright (c) Squidex Group
+//  All rights reserved.
+// ==========================================================================
+namespace Squidex.Domain.Apps.Core.Apps
+{
+    public class AppPattern
```
Comment (SebastianStehle) [inline src/Squidex.Domain.Apps.Core.Model/Apps/AppPattern.cs]: Classes must be immutable in the Core Model.

File: src/Squidex.Domain.Apps.Core.Model/Apps/AppPatterns.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+//  Copyright (c) Squidex Group
+//  All rights reserved.
+// ==========================================================================
+using System.Collections.Immutable;
+using System.Diagnostics.Contracts;
+using Squidex.Infrastructure;
+
+namespace Squidex.Domain.Apps.Core.Apps
+{
+    public sealed class AppPatterns : DictionaryWrapper<string, AppPattern>
```
Comment (SebastianStehle) [inline src/Squidex.Domain.Apps.Core.Model/Apps/AppPatterns.cs]: Classes must be immutable.

File: src/Squidex.Domain.Apps.Write/Apps/Commands/UpdatePattern.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+using System;
+using System.Collections.Generic;
+using Squidex.Domain.Apps.Core.Schemas;
+using Squidex.Infrastructure;
+
+namespace Squidex.Domain.Apps.Write.Apps.Commands
+{
+    public sealed class UpdatePattern : AppAggregateCommand, IValidatable
+    {
+        public string OriginalName { get; set; }
```
Comment (SebastianStehle) [inline src/Squidex.Domain.Apps.Write/Apps/Commands/UpdatePattern.cs]: Why do I need the original values?

File: src/Squidex/Areas/Api/Controllers/Apps/AppPatternsController.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        /// <remarks>
+        /// Gets all configured regex patterns for the app with the specified name.
+        /// </remarks>
+        [HttpGet]
+        [Route("apps/{app}/patterns/")]
+        [ProducesResponseType(typeof(AppPatternDto[]), 200)]
+        [ApiCosts(1)]
+        public IActionResult GetPatterns(string app)
+        {
+            return Ok(App.Patterns.Values?
```
Comment (SebastianStehle) [inline src/Squidex/Areas/Api/Controllers/Apps/AppPatternsController.cs]: A temporary variable would make debugging much easier.

File: src/Squidex/Areas/Api/Controllers/Apps/Models/AppPatternDto.cs [hunk 5]
```diff
@@ -1,16 +1,15 @@
...
+//  AppPatternDto.cs
 //  Squidex Headless CMS
 // ==========================================================================
 //  Copyright (c) Squidex Group
 //  All rights reserved.
 // ==========================================================================
-
 using System.ComponentModel.DataAnnotations;
 
 namespace Squidex.Areas.Api.Controllers.UI.Models
```
Comment (SebastianStehle) [inline src/Squidex/Areas/Api/Controllers/Apps/Models/AppPatternDto.cs]: Wrong namespace

File: src/Squidex/Areas/Api/Controllers/Apps/AppPatternsController.cs (hunk 4 above)
Comment (SebastianStehle) [reply src/Squidex/Areas/Api/Controllers/Apps/AppPatternsController.cs]: Can we make a fallback here to the config, when the pattern list is null? Would prevent us from having migrations.

File: src/Squidex/Areas/Api/Controllers/Apps/AppPatternsController.cs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+using Squidex.Infrastructure.Reflection;
+using Squidex.Pipeline;
+
+namespace Squidex.Areas.Api.Controllers.Apps
+{
+    /// <summary>
+    /// Manages and configures app patterns.
+    /// </summary>
+    [ApiAuthorize]
+    [MustBeAppOwner]
```
Comment (SebastianStehle) [inline src/Squidex/Areas/Api/Controllers/Apps/AppPatternsController.cs]: I think developer should be enough

File: src/Squidex/app/features/settings/settings-area.component.html [hunk 7]
```diff
@@ -32,12 +32,19 @@ <h3 class="panel-title">Settings</h3>
                         <i class="icon-angle-right"></i>
                     </a>
                 </li>
+                <li class="nav-item">
```
Comment (SebastianStehle) [inline src/Squidex/app/features/settings/settings-area.component.html]: Please check permissions here

File: src/Squidex.Domain.Apps.Write/Apps/Guards/GuardAppPattern.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                {
+                    error(new ValidationError("Pattern can not be empty.", "Name"));
+                }
+
+                if (!patterns.ContainsKey(command.OriginalName.ToLower()))
+                {
+                    error(new ValidationError("Pattern not found.", nameof(command.OriginalName)));
+                }
+
+                if (patterns.Where(x => !x.Key.Equals(command.OriginalName, StringComparison.OrdinalIgnoreCase))
```
Comment (SebastianStehle) [inline src/Squidex.Domain.Apps.Write/Apps/Guards/GuardAppPattern.cs]: I dont understand this line. Why do you need a dictionary when you have alreaded looped over all the items?

File: src/Squidex.Domain.Apps.Write/Apps/Guards/GuardAppPattern.cs [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+                    .ToDictionary(x => x.Key, y => y.Value).ContainsKey(command.Name.ToLower()))
+                {
+                    error(new ValidationError("Pattern name is already assigned.", "Name"));
+                }
+
+                if (patterns.Where(x => !x.Key.Equals(command.OriginalName, StringComparison.OrdinalIgnoreCase))
```
Comment (SebastianStehle) [inline src/Squidex.Domain.Apps.Write/Apps/Guards/GuardAppPattern.cs]: I dont understand this line. Why do you need a dictionary when you have alreaded looped over all the items?

**Comment by dsbegnoche** [comment]
1 - Agreed, I don't recall the though process behind using name as keys.  Caused alot of extra code and is unclear.  Will add Id's to use as Keys
2 - Are you saying that instead of updating/modifying/deleting a single pattern, instead we would update the entire collection at once?
3 - Sorry about that.  I recently started using VSCode, and apparently have some formatting problems with it.  I do not seem to have them with VS itself, so need to figure out what is going on there.  I forgot to go back and fix the formatting before pushing :(

**Comment by SebastianStehle** [comment]
1. Depends on 2.
2. Yes, I mean, how many items are there. We talk about very small payloads.
3. No problem, just recognized it.

**Comment by dsbegnoche** [comment]
Added Id for Key, and cleaned up some code.  As for saving the all patterns at once, I am not sure on doing that.  One thing I mentioned in initial PR, was that we are looking to update any field that is using a pattern when the pattern is updated.  First, I am not sure on best way to handle that, but I think that saving the entire list at once will make that more difficult to handle.

**Comment by SebastianStehle** [comment]
Looks very good for me :) ... I am working on the CQRS stuff at the moment and will it merge it after that. For 100% next week.

**Comment by dsbegnoche** [comment]
Thanks Sebastian.  Do you have any thoughts on how we could solve the update requirement on our end?  Right now it just loops through all schemas to find files using the pattern, and updates them.  I don't like that method though.  My only other thought right now though is to maybe have mongo do it with a filter and update?

**Comment by SebastianStehle** [comment]
The fastest approach would be to use  a secondary index

**Comment by SebastianStehle** [comment]
Hi @dsbegnoche 
 
Because there are so many conflicts I have integrated the code step by step by copying files and checking them. Also found some bugs ;)

https://github.com/Squidex/squidex/pull/208

**Pull request closed without merging**